runtime/drivers/salesforce/authentication.go

```diff
@@ -0,0 +1,109 @@
+package salesforce
+
+import (
+	"errors"
+	"fmt"
+	"net/url"
+
+	force "github.com/ForceCLI/force/lib"
+)
+
+const defaultEndpoint = "https://login.salesforce.com"
+
+type authenticationOptions struct {
+	Endpoint     string
+	Username     string
+	Password     string
+	JWT          string
+	ConnectedApp string
+}
+
+// Authenticable represents the authentation subset of the force api
+type Authenticable interface {
+	JwtAssertionForEndpoint(endpoint string, username string, keyfile string, clientID string) (string, error)
+	JwtLoginAtEndpoint(endpoint string, assertion string) (*force.Force, error)
+	ForceSoapLoginAtEndpoint(endpoint string, userName string, password string) (*force.Force, error)
+}
+
+// a concrete implementation of Authenticable
+type forceProvider struct{}
+
+func authenticate(options authenticationOptions) (*force.Force, error) {
+	if options.ConnectedApp == "" {
+		options.ConnectedApp = defaultClientID
+	}
+	force.ClientId = options.ConnectedApp
+	provider := forceProvider{}
+
+	isUsernameSelected := len(options.Username) > 0
+	isJWTSelected := len(options.JWT) > 0
+	isSOAPSelected := len(options.Password) > 0
+
+	endpoint, err := endpoint(options)
+	if err != nil {
+		return nil, err
+	}
+
+	switch {
+	case isJWTSelected:
+		return jwtLogin(endpoint, options, provider)
+	case isSOAPSelected:
+		return provider.ForceSoapLoginAtEndpoint(endpoint, options.Username, options.Password)
+	case !isUsernameSelected:
+		return nil, fmt.Errorf("username missing")
+	}
+	return nil, fmt.Errorf("unable to authenticate")
+}
+
+func endpoint(options authenticationOptions) (endpoint string, err error) {
+	isEndpointSelected := len(options.Endpoint) > 0
+
+	if isEndpointSelected {
+		// URL needs to have scheme lest the force cli lib chokes
+		uri, err := url.Parse(options.Endpoint)
+		if err != nil {
+			return defaultEndpoint, errors.New("Unable to parse endpoint: " + options.Endpoint)
+		}
+
+		if uri.Host == "" {
```

Comment:
Shouldn't this be `uri.Schema == ""` ?

Reply:
Also can be simplified to 
```
if u.Scheme == "" {
	u.Scheme = "https"
}
```

Reply:
Added tests in a2446ca17a34da31fb0876cbbc49eb7da67ee26f to illustrate support for when `authenticationOptions.Endpoint` contains a hostname rather than a URL.

Reply:
I misunderstood how `URL` works.  I didn't realize that `URL.Host` could be the empty string, and `URL.String()` would still return the hostname (or the URL after `URL.Scheme` is set).

Updated in 4b048742431412d67003cc0c4fd85a55ca39d9f4.